Clicking email confirmation links in Outlook results in a "stale link" error

### Before reporting an issue

- [x] I have read and understood the above terms for submitting issues, and I understand that my issue may be closed without action if I do not follow them.

### Area

core

### Describe the bug

When using the UPDATE_EMAIL feature, updating your email address in the account console and then clicking on the verification link in some email clients (at least Outlook) results in an error:
<img width="578" height="149" alt="Image" src="https://github.com/user-attachments/assets/9c5a0d86-5e33-4551-a3c7-39b36bad0166" />

The email address is still successfully updated, but the error message might be confusing for the user.

---

The issue is caused by Outlook "Safe Links" [sending a HEAD request](https://community.wappler.io/t/dealing-with-head-requests/33802), which already verifies the email address. The GET request from the browser then results in a EXECUTE_ACTIONS_ERROR with reason "staleEmailVerificationLink".

The verification link upon initial registration also has this issue, but in that case doesn't result in a user-visible error, because the VERIFY_EMAIL_ERROR with error "email_already_verified" still results in an automatic redirect back to the application, just as a successful VERIFY_EMAIL would.

### Version

26.2.5

### Regression

- [ ] The issue is a regression

### Expected behavior

HEAD requests to the `/realms/<realm>/login-actions/action-token` endpoint should not cause any state change on the server.

### Actual behavior

HEAD requests to the `/realms/<realm>/login-actions/action-token` endpoint complete the action and invalidate the token.

### How to Reproduce?

1. Enable the UPDATE_EMAIL feature and the corresponding "Update Email" required action.
2. Open the account console and log in to an existing account. Click the "Update email address" button
3. Enter a valid email address and click "Submit"
4. Open the email in Microsoft Outlook and click the link. (This only works with publicly accessible Keycloak instances on the internet, as the HEAD request originates from Microsoft's servers).  
If you don't have Outlook or your Keycloak instance isn't public, you can instead copy the verification link and manually send a HEAD request using a tool of your choice, e.g. `curl -I <url>`, and then open the link in your browser.
5. You will see "stale link" error message.

### Anything else?

Keycloak event log:
<img width="1389" height="111" alt="Image" src="https://github.com/user-attachments/assets/1846ce51-5535-4900-90a3-43d5ee946b3f" />

nginx log:
<img width="1890" height="487" alt="Image" src="https://github.com/user-attachments/assets/96573903-24ce-4561-af25-ecc894039d0f" />

You can see the HEAD request from a Microsoft IP arriving first and getting a 200 response, then the GET request from my browser arriving second and getting a 400 response.

---

The responsible code should be this: https://github.com/keycloak/keycloak/blob/ef312b570c2ed0243242b93688016d7aaf6d91fd/services/src/main/java/org/keycloak/services/resources/LoginActionsService.java#L532-L549

It seems that if no dedicated `@HEAD` handler is defined, the `@GET` handler is called instead (which in this case has side effects).

There is an argument to be made that neither HEAD nor GET requests should actually cause an action to complete, because that violates the assumption of them being [safe](https://developer.mozilla.org/en-US/docs/Glossary/Safe/HTTP) and [idempotent](https://developer.mozilla.org/en-US/docs/Glossary/Idempotent). But to solve that, the verification link would have to always show a form with a button that sends a POST request. To solve this issue, it would be enough for HEAD requests to not execute the action token handler.

> [@stianst](https://github.com/stianst) Yes, I did use `-DskipTests` and even tried skipping individual failing build steps. I think the problem was that proxy settings (as set by `http_proxy` and `https_proxy` environment variables) weren't being respected at multiple steps. At some point I just gave up.
> 
> Trying it now, this is the error I get when running `./mvnw clean install -DskipTests -e`:
> 
> Error
> 
> ```
> [ERROR] Failed to execute goal org.infinispan.protostream:proto-schema-compatibility-maven-plugin:5.0.14.Final:proto-schema-compatibility-check (default) on project keycloak-model-infinispan: An error occurred while running protolock: Connection timed out -> [Help 1]
> org.apache.maven.lifecycle.LifecycleExecutionException: Failed to execute goal org.infinispan.protostream:proto-schema-compatibility-maven-plugin:5.0.14.Final:proto-schema-compatibility-check (default) on project keycloak-model-infinispan: An error occurred while running protolock
>     at org.apache.maven.lifecycle.internal.MojoExecutor.doExecute2 (MojoExecutor.java:333)
>     at org.apache.maven.lifecycle.internal.MojoExecutor.doExecute (MojoExecutor.java:316)
>     at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:212)
>     at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:174)
>     at org.apache.maven.lifecycle.internal.MojoExecutor.access$000 (MojoExecutor.java:75)
>     at org.apache.maven.lifecycle.internal.MojoExecutor$1.run (MojoExecutor.java:162)
>     at org.apache.maven.buildcache.BuildCacheMojosExecutionStrategy.execute (BuildCacheMojosExecutionStrategy.java:145)
>     at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:159)
>     at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:105)
>     at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:73)
>     at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:53)
>     at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:118)
>     at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:261)
>     at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:173)
>     at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:101)
>     at org.apache.maven.cli.MavenCli.execute (MavenCli.java:903)
>     at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:280)
>     at org.apache.maven.cli.MavenCli.main (MavenCli.java:203)
>     at jdk.internal.reflect.DirectMethodHandleAccessor.invoke (DirectMethodHandleAccessor.java:103)
>     at java.lang.reflect.Method.invoke (Method.java:580)
>     at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:255)
>     at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:201)
>     at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:361)
>     at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:314)
> Caused by: org.apache.maven.plugin.MojoExecutionException: An error occurred while running protolock
>     at org.infinispan.plugins.proto.compatibility.ProtoCompatibilityMojo.execute (ProtoCompatibilityMojo.java:99)
>     at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo (DefaultBuildPluginManager.java:126)
>     at org.apache.maven.lifecycle.internal.MojoExecutor.doExecute2 (MojoExecutor.java:328)
>     at org.apache.maven.lifecycle.internal.MojoExecutor.doExecute (MojoExecutor.java:316)
>     at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:212)
>     at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:174)
>     at org.apache.maven.lifecycle.internal.MojoExecutor.access$000 (MojoExecutor.java:75)
>     at org.apache.maven.lifecycle.internal.MojoExecutor$1.run (MojoExecutor.java:162)
>     at org.apache.maven.buildcache.BuildCacheMojosExecutionStrategy.execute (BuildCacheMojosExecutionStrategy.java:145)
>     at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:159)
>     at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:105)
>     at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:73)
>     at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:53)
>     at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:118)
>     at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:261)
>     at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:173)
>     at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:101)
>     at org.apache.maven.cli.MavenCli.execute (MavenCli.java:903)
>     at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:280)
>     at org.apache.maven.cli.MavenCli.main (MavenCli.java:203)
>     at jdk.internal.reflect.DirectMethodHandleAccessor.invoke (DirectMethodHandleAccessor.java:103)
>     at java.lang.reflect.Method.invoke (Method.java:580)
>     at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:255)
>     at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:201)
>     at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:361)
>     at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:314)
> Caused by: java.net.ConnectException: Connection timed out
>     at sun.nio.ch.Net.connect0 (Native Method)
>     at sun.nio.ch.Net.connect (Net.java:589)
>     at sun.nio.ch.Net.connect (Net.java:578)
>     at sun.nio.ch.NioSocketImpl.connect (NioSocketImpl.java:583)
>     at java.net.SocksSocketImpl.connect (SocksSocketImpl.java:327)
>     at java.net.Socket.connect (Socket.java:751)
>     at sun.security.ssl.SSLSocketImpl.connect (SSLSocketImpl.java:304)
>     at sun.security.ssl.BaseSSLSocketImpl.connect (BaseSSLSocketImpl.java:181)
>     at sun.net.NetworkClient.doConnect (NetworkClient.java:183)
>     at sun.net.www.http.HttpClient.openServer (HttpClient.java:531)
>     at sun.net.www.http.HttpClient.openServer (HttpClient.java:636)
>     at sun.net.www.protocol.https.HttpsClient.<init> (HttpsClient.java:264)
>     at sun.net.www.protocol.https.HttpsClient.New (HttpsClient.java:377)
>     at sun.net.www.protocol.https.AbstractDelegateHttpsURLConnection.getNewHttpClient (AbstractDelegateHttpsURLConnection.java:193)
>     at sun.net.www.protocol.http.HttpURLConnection.plainConnect0 (HttpURLConnection.java:1253)
>     at sun.net.www.protocol.http.HttpURLConnection.plainConnect (HttpURLConnection.java:1139)
>     at sun.net.www.protocol.https.AbstractDelegateHttpsURLConnection.connect (AbstractDelegateHttpsURLConnection.java:179)
>     at sun.net.www.protocol.http.HttpURLConnection.getInputStream0 (HttpURLConnection.java:1691)
>     at sun.net.www.protocol.http.HttpURLConnection.getInputStream (HttpURLConnection.java:1615)
>     at sun.net.www.protocol.https.HttpsURLConnectionImpl.getInputStream (HttpsURLConnectionImpl.java:223)
>     at java.net.URL.openStream (URL.java:1325)
>     at org.infinispan.plugins.proto.compatibility.ProtoCompatibilityMojo.checkRemoteCompatibility (ProtoCompatibilityMojo.java:125)
>     at org.infinispan.plugins.proto.compatibility.ProtoCompatibilityMojo.execute (ProtoCompatibilityMojo.java:86)
>     at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo (DefaultBuildPluginManager.java:126)
>     at org.apache.maven.lifecycle.internal.MojoExecutor.doExecute2 (MojoExecutor.java:328)
>     at org.apache.maven.lifecycle.internal.MojoExecutor.doExecute (MojoExecutor.java:316)
>     at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:212)
>     at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:174)
>     at org.apache.maven.lifecycle.internal.MojoExecutor.access$000 (MojoExecutor.java:75)
>     at org.apache.maven.lifecycle.internal.MojoExecutor$1.run (MojoExecutor.java:162)
>     at org.apache.maven.buildcache.BuildCacheMojosExecutionStrategy.execute (BuildCacheMojosExecutionStrategy.java:145)
>     at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:159)
>     at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:105)
>     at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:73)
>     at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:53)
>     at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:118)
>     at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:261)
>     at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:173)
>     at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:101)
>     at org.apache.maven.cli.MavenCli.execute (MavenCli.java:903)
>     at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:280)
>     at org.apache.maven.cli.MavenCli.main (MavenCli.java:203)
>     at jdk.internal.reflect.DirectMethodHandleAccessor.invoke (DirectMethodHandleAccessor.java:103)
>     at java.lang.reflect.Method.invoke (Method.java:580)
>     at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:255)
>     at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:201)
>     at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:361)
>     at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:314)
> [ERROR] 
> [ERROR] Re-run Maven using the -X switch to enable full debug logging.
> [ERROR] 
> [ERROR] For more information about the errors and possible solutions, please read the following articles:
> [ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoExecutionException
> ```
> 
> But this probably belongs elsewhere and not these issue comments.

@ryanemerson seems the proto stuff doesn't honor proxy settings?

--- a/services/src/main/java/org/keycloak/headers/DefaultSecurityHeadersProvider.java
+++ b/services/src/main/java/org/keycloak/headers/DefaultSecurityHeadersProvider.java
@@ -30,6 +30,8 @@ import jakarta.ws.rs.core.MultivaluedMap;
 import java.util.Collections;
 import java.util.Map;
 
+import static jakarta.ws.rs.HttpMethod.HEAD;
+import static jakarta.ws.rs.HttpMethod.OPTIONS;
 import static org.keycloak.models.BrowserSecurityHeaders.CONTENT_SECURITY_POLICY;
 
 public class DefaultSecurityHeadersProvider implements SecurityHeadersProvider {
@@ -147,10 +149,17 @@ public class DefaultSecurityHeadersProvider implements SecurityHeadersProvider {
                 status == 400 || status == 401 || status == 403 || status == 404) {
                 return true;
             }
-            if (requestContext.getMethod().equalsIgnoreCase("OPTIONS")) {
-                return true;
+
+            String method = requestContext.getMethod().toUpperCase();
+
+            switch (method) {
+                case OPTIONS:
+                    return true;
+                case HEAD:
+                    return status == 200;
             }
         }
+
         return false;
     }
 

--- a/services/src/main/java/org/keycloak/services/resources/LoginActionsService.java
+++ b/services/src/main/java/org/keycloak/services/resources/LoginActionsService.java
@@ -16,6 +16,7 @@
  */
 package org.keycloak.services.resources;
 
+import jakarta.ws.rs.HEAD;
 import org.jboss.logging.Logger;
 import org.keycloak.common.Profile;
 import org.keycloak.common.Profile.Feature;
@@ -546,6 +547,19 @@ public class LoginActionsService {
                                        @QueryParam(Constants.CLIENT_DATA) String clientData,
                                        @QueryParam(Constants.TAB_ID) String tabId) {
         return handleActionToken(key, execution, clientId, tabId, clientData, null);
+    }
+
+    /**
+     * Skip processing {@link jakarta.ws.rs.HttpMethod#HEAD} requests for action tokens
+     * as they are usually used by mail servers to validate links. The actual request will eventually be
+     * processed by the {@link #executeActionToken} method.
+     *
+     * @return a {@link Response.Status#OK} response with no message body
+     */
+    @Path("action-token")
+    @HEAD
+    public Response executeActionTokenHead() {
+        return Response.ok().build();
     }
 
     protected <T extends JsonWebToken & SingleUseObjectKeyModel> Response handleActionToken(String tokenString, String execution, String clientId, String tabId, String clientData, 

--- a/testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/actions/RequiredActionUpdateEmailTestWithVerificationTest.java
+++ b/testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/actions/RequiredActionUpdateEmailTestWithVerificationTest.java
@@ -30,8 +30,10 @@ import java.io.IOException;
 import java.util.List;
 import java.util.UUID;
 
+import jakarta.ws.rs.core.Response.Status;
+import org.apache.http.impl.client.CloseableHttpClient;
+import org.apache.http.impl.client.HttpClientBuilder;
 import org.jboss.arquillian.graphene.page.Page;
-import org.jetbrains.annotations.NotNull;
 import org.junit.Assert;
 import org.junit.Rule;
 import org.junit.Test;
@@ -39,6 +41,7 @@ import org.keycloak.admin.client.resource.AuthenticationManagementResource;
 import org.keycloak.admin.client.resource.UserResource;
 import org.keycloak.authentication.actiontoken.updateemail.UpdateEmailActionToken;
 import org.keycloak.authentication.requiredactions.UpdateEmail;
+import org.keycloak.broker.provider.util.SimpleHttp.Response;
 import org.keycloak.events.Details;
 import org.keycloak.events.EventType;
 import org.keycloak.models.UserModel;
@@ -48,6 +51,7 @@ import org.keycloak.representations.idm.RealmRepresentation;
 import org.keycloak.representations.idm.RequiredActionProviderRepresentation;
 import org.keycloak.representations.idm.UserRepresentation;
 import org.keycloak.representations.idm.UserSessionRepresentation;
+import org.keycloak.testsuite.broker.util.SimpleHttpDefault;
 import org.keycloak.testsuite.pages.ErrorPage;
 import org.keycloak.testsuite.pages.InfoPage;
 import org.keycloak.testsuite.util.GreenMailRule;
@@ -173,6 +177,26 @@ public class RequiredActionUpdateEmailTestWithVerificationTest extends AbstractR
 		assertEquals("The link you clicked is an old stale link and is no longer valid. Maybe you have already verified your email.", errorPage.getError());
 		assertTrue(ActionUtil.findUserWithAdminClient(adminClient, "test-user@localhost").getRequiredActions().contains(UserModel.RequiredAction.UPDATE_EMAIL.name()));
 	}
+
+    @Test
+    public void testSkipHeadRequestWhenFollowingVerificationLink() throws MessagingException, IOException {
+        oauth.openLoginForm();
+        loginPage.login("test-user@localhost", "password");
+
+        updateEmailPage.assertCurrent();
+        updateEmailPage.changeEmail("new@localhost");
+
+        String confirmationLink = fetchEmailConfirmationLink("new@localhost");
+
+        try (CloseableHttpClient httpClient = HttpClientBuilder.create().build()) {
+            try (Response response = SimpleHttpDefault.doHead(confirmationLink, httpClient).asResponse()) {
+                assertEquals(Status.OK.getStatusCode(), response.getStatus());
+            }
+        }
+
+        driver.navigate().to(confirmationLink);
+        infoPage.assertCurrent();
+    }
 
     @Test
     public void testForceEmailVerification() throws MessagingException, IOException {

